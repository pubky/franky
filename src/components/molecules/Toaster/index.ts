export * from './Toaster';
export * from './use-toast';
