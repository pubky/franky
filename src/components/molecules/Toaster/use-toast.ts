'use client';

import { toast as sonnerToast } from 'sonner';
import * as React from 'react';

type ToastVariant = 'default' | 'error' | 'warning' | 'success';

type Toast = {
  title?: React.ReactNode;
  description?: React.ReactNode;
  variant?: ToastVariant;
  action?: {
    label: string;
    onClick: () => void;
  };
};

function toast({ title, description, action, ...props }: Toast) {
  return sonnerToast(title, {
    description: description,
    action: action,
    ...props,
  });
}

function useToast() {
  return {
    toast,
    dismiss: (toastId?: string) => {
      if (toastId) {
        sonnerToast.dismiss(toastId);
      } else {
        sonnerToast.dismiss();
      }
    },
  };
}

export { useToast, toast };
