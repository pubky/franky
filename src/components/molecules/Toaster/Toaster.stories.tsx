import type { Meta, StoryObj } from '@storybook/nextjs-vite';

import { Toaster } from './Toaster';
import { Button } from '@/components/atoms/Button';
import { useToast } from './use-toast';

const meta: Meta<typeof Toaster> = {
  title: 'Molecules/Toaster',
  component: Toaster,
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component:
          'The Toaster component provides a centralized toast notification system. It renders a ToastViewport that displays toast notifications at the bottom-center of the screen with proper theming and styling.',
      },
    },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

const InteractiveExample = () => {
  const { toast } = useToast();

  return (
    <div className="max-w-md mx-auto space-y-4">
      <h2 className="text-2xl font-bold">Interactive Toaster</h2>
      <p className="text-muted-foreground">Click the buttons below to see different types of toast notifications.</p>
      <div className="grid grid-cols-1 gap-2">
        <Button
          onClick={() =>
            toast({
              title: 'Success!',
              description: 'Your action was completed successfully.',
            })
          }
        >
          Show Success Toast
        </Button>
        <Button
          variant="destructive"
          onClick={() =>
            toast({
              title: 'Error!',
              description: 'Something went wrong. Please try again.',
              variant: 'error',
            })
          }
        >
          Show Error Toast
        </Button>
        <Button
          variant="outline"
          onClick={() =>
            toast({
              title: 'Warning!',
              description: 'Please review your input before continuing.',
              variant: 'warning',
            })
          }
        >
          Show Warning Toast
        </Button>
      </div>
    </div>
  );
};

export const Default: Story = {
  render: () => (
    <div className="min-h-screen bg-background p-8">
      <InteractiveExample />
      <Toaster />
    </div>
  ),
};

export const WithBackground: Story = {
  render: () => (
    <div className="min-h-screen bg-gradient-to-br from-blue-50 to-indigo-100 p-8">
      <div className="max-w-md mx-auto space-y-4">
        <h2 className="text-2xl font-bold text-gray-900">Toaster with Background</h2>
        <p className="text-gray-700">The toaster works with different background colors and themes.</p>
        <div className="bg-white p-4 rounded-lg shadow-sm">
          <p className="text-sm text-gray-600 mb-4">This is a content area where toasts would appear above.</p>
          <InteractiveExample />
        </div>
      </div>
      <Toaster />
    </div>
  ),
};

export const DarkTheme: Story = {
  render: () => (
    <div className="min-h-screen bg-gray-900 text-white p-8">
      <div className="max-w-md mx-auto space-y-4">
        <h2 className="text-2xl font-bold">Toaster in Dark Theme</h2>
        <p className="text-gray-300">The toaster adapts to dark themes automatically.</p>
        <div className="bg-gray-800 p-4 rounded-lg border border-gray-700">
          <p className="text-sm text-gray-400 mb-4">
            Toast notifications will appear with appropriate dark theme styling.
          </p>
          <InteractiveExample />
        </div>
      </div>
      <Toaster />
    </div>
  ),
};

const ToastWithActionsExample = () => {
  const { toast } = useToast();

  return (
    <div className="max-w-md mx-auto space-y-4">
      <h3 className="text-lg font-semibold">Toast with Actions</h3>
      <div className="grid grid-cols-1 gap-2">
        <Button
          onClick={() =>
            toast({
              title: 'File uploaded',
              description: 'Your file has been successfully uploaded.',
              action: {
                label: 'View',
                onClick: () => console.log('View file'),
              },
            })
          }
        >
          Upload File
        </Button>
        <Button
          onClick={() =>
            toast({
              title: 'Changes saved',
              description: 'Your changes have been saved to the database.',
              action: {
                label: 'Undo',
                onClick: () => console.log('Undo changes'),
              },
            })
          }
        >
          Save Changes
        </Button>
      </div>
    </div>
  );
};

export const WithActions: Story = {
  render: () => (
    <div className="min-h-screen bg-background p-8">
      <ToastWithActionsExample />
      <Toaster />
    </div>
  ),
};

const MultipleToastsExample = () => {
  const { toast } = useToast();

  const showMultipleToasts = () => {
    toast({
      title: 'First Toast',
      description: 'This is the first notification.',
    });

    setTimeout(() => {
      toast({
        title: 'Second Toast',
        description: 'This is the second notification.',
        variant: 'warning',
      });
    }, 500);

    setTimeout(() => {
      toast({
        title: 'Third Toast',
        description: 'This is the third notification.',
        variant: 'error',
      });
    }, 1000);
  };

  return (
    <div className="max-w-md mx-auto space-y-4">
      <h3 className="text-lg font-semibold">Multiple Toasts</h3>
      <p className="text-sm text-muted-foreground">
        Click the button below to show multiple toasts with different variants.
      </p>
      <Button onClick={showMultipleToasts}>Show Multiple Toasts</Button>
    </div>
  );
};

export const MultipleToasts: Story = {
  render: () => (
    <div className="min-h-screen bg-background p-8">
      <MultipleToastsExample />
      <Toaster />
    </div>
  ),
};

const DismissibleToastExample = () => {
  const { toast } = useToast();

  return (
    <div className="max-w-md mx-auto space-y-4">
      <h3 className="text-lg font-semibold">Dismissible Toast</h3>
      <p className="text-sm text-muted-foreground">
        This toast includes an &quot;OK&quot; button that dismisses the toast when clicked.
      </p>
      <div className="grid grid-cols-1 gap-2">
        <Button
          onClick={() =>
            toast({
              title: 'Information',
              description: 'This is an important message that requires acknowledgment.',
              action: {
                label: 'OK',
                onClick: () => {
                  // The toast will be dismissed automatically when the action is clicked
                  console.log('Toast dismissed');
                },
              },
            })
          }
        >
          Show Dismissible Toast
        </Button>
        <Button
          variant="destructive"
          onClick={() =>
            toast({
              title: 'Error Occurred',
              description: 'Something went wrong. Please review and try again.',
              action: {
                label: 'Dismiss',
                onClick: () => {
                  console.log('Error toast dismissed');
                },
              },
            })
          }
        >
          Show Error with Dismiss
        </Button>
        <Button
          variant="outline"
          onClick={() =>
            toast({
              title: 'Warning',
              description: 'Please review your settings before continuing.',
              action: {
                label: 'Got it',
                onClick: () => {
                  console.log('Warning acknowledged');
                },
              },
            })
          }
        >
          Show Warning with Acknowledge
        </Button>
      </div>
    </div>
  );
};

export const DismissibleToast: Story = {
  render: () => (
    <div className="min-h-screen bg-background p-8">
      <DismissibleToastExample />
      <Toaster />
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Demonstrates toasts with dismissible action buttons, similar to real app notifications.',
      },
    },
  },
};

const RealWorldToastExample = () => {
  const { toast } = useToast();

  return (
    <div className="max-w-md mx-auto space-y-4">
      <h3 className="text-lg font-semibold">Real-World Toast Examples</h3>
      <p className="text-sm text-muted-foreground">
        Examples of toasts you might see in a real application with proper dismiss buttons.
      </p>
      <div className="grid grid-cols-1 gap-2">
        <Button
          onClick={() =>
            toast({
              title: 'Settings Saved',
              description: 'Your preferences have been updated successfully.',
              action: {
                label: 'OK',
                onClick: () => {
                  console.log('Settings toast acknowledged');
                },
              },
            })
          }
        >
          Settings Saved Toast
        </Button>
        <Button
          variant="destructive"
          onClick={() =>
            toast({
              title: 'Connection Lost',
              description: 'Unable to connect to the server. Please check your internet connection.',
              action: {
                label: 'Retry',
                onClick: () => {
                  console.log('Retrying connection...');
                  // In a real app, this would trigger a reconnection attempt
                },
              },
            })
          }
        >
          Connection Error Toast
        </Button>
        <Button
          variant="outline"
          onClick={() =>
            toast({
              title: 'New Message',
              description: 'You have received a new message from John Doe.',
              action: {
                label: 'View',
                onClick: () => {
                  console.log('Opening message...');
                  // In a real app, this would navigate to the message
                },
              },
            })
          }
        >
          New Message Toast
        </Button>
        <Button
          onClick={() =>
            toast({
              title: 'File Upload Complete',
              description: 'Your file has been uploaded and is ready for processing.',
              action: {
                label: 'View File',
                onClick: () => {
                  console.log('Opening uploaded file...');
                },
              },
            })
          }
        >
          File Upload Toast
        </Button>
      </div>
    </div>
  );
};

export const RealWorldExamples: Story = {
  render: () => (
    <div className="min-h-screen bg-background p-8">
      <RealWorldToastExample />
      <Toaster />
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Real-world examples of toasts with dismissible action buttons, similar to what users see in production applications.',
      },
    },
  },
};
