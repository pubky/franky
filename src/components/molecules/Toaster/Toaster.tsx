'use client';

import { Toast, ToastDescription, ToastProvider, ToastTitle, ToastViewport } from '@/atoms';
import { useToast } from './use-toast';

export function Toaster() {
  const { toasts } = useToast();

  return (
    <ToastProvider>
      {toasts.map(function ({ id, title, description, action, variant, ...props }) {
        return (
          <Toast
            key={id}
            variant={variant}
            {...props}
          >
            <div className="flex items-center justify-between gap-4 w-full">
              <div className="flex flex-col gap-1 w-full">
                {title && <ToastTitle>{title}</ToastTitle>}
                {description && <ToastDescription>{description}</ToastDescription>}
              </div>
              <div className="flex items-center justify-between gap-4">{action}</div>
            </div>
          </Toast>
        );
      })}
      <ToastViewport />
    </ToastProvider>
  );
}
