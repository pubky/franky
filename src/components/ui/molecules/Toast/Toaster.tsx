'use client';

import { Toast, ToastDescription, ToastProvider, ToastTitle, ToastViewport } from './Toast';
import { useToast } from './use-toast';

export function Toaster() {
  const { toasts } = useToast();

  return (
    <ToastProvider>
      {toasts.map(function ({ id, title, description, action, ...props }) {
        return (
          <Toast
            key={id}
            className="flex items-center justify-between gap-4 p-6 rounded-lg bg-black-900 border border border-brand/20 text-white"
            {...props}
          >
            <div className="flex items-center justify-between gap-4 w-full">
              <div className="flex flex-col gap-1 w-full">
                {title && <ToastTitle>{title}</ToastTitle>}
                {description && <ToastDescription>{description}</ToastDescription>}
              </div>
              <div className="flex items-center justify-between gap-4">{action}</div>
            </div>
          </Toast>
        );
      })}
      <ToastViewport />
    </ToastProvider>
  );
}
