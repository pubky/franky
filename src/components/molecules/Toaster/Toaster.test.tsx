import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { Toaster } from './Toaster';

// Mock next-themes
vi.mock('next-themes', () => ({
  useTheme: vi.fn(),
}));

// Mock sonner
vi.mock('sonner', () => ({
  Toaster: ({ children, ...props }: { children: React.ReactNode; [key: string]: unknown }) => {
    return (
      <div data-testid="sonner-toaster" {...props}>
        {children}
      </div>
    );
  },
}));

// Mock @/libs to intercept any icons and utilities
vi.mock('@/libs', () => ({
  X: () => <svg data-testid="x-icon" />,
  Radio: ({ className }: { className?: string }) => (
    <div data-testid="radio-icon" className={className}>
      Radio
    </div>
  ),
  UsersRound2: ({ className }: { className?: string }) => (
    <div data-testid="users-round2-icon" className={className}>
      UsersRound2
    </div>
  ),
  HeartHandshake: ({ className }: { className?: string }) => (
    <div data-testid="heart-handshake-icon" className={className}>
      HeartHandshake
    </div>
  ),
  UserRound: ({ className }: { className?: string }) => (
    <div data-testid="user-round-icon" className={className}>
      UserRound
    </div>
  ),
  SquareAsterisk: ({ className }: { className?: string }) => (
    <div data-testid="square-asterisk-icon" className={className}>
      SquareAsterisk
    </div>
  ),
  Flame: ({ className }: { className?: string }) => (
    <div data-testid="flame-icon" className={className}>
      Flame
    </div>
  ),
  Columns3: ({ className }: { className?: string }) => (
    <div data-testid="columns3-icon" className={className}>
      Columns3
    </div>
  ),
  Menu: ({ className }: { className?: string }) => (
    <div data-testid="menu-icon" className={className}>
      Menu
    </div>
  ),
  LayoutGrid: ({ className }: { className?: string }) => (
    <div data-testid="layout-grid-icon" className={className}>
      LayoutGrid
    </div>
  ),
  Layers: ({ className }: { className?: string }) => (
    <div data-testid="layers-icon" className={className}>
      Layers
    </div>
  ),
  StickyNote: ({ className }: { className?: string }) => (
    <div data-testid="sticky-note-icon" className={className}>
      StickyNote
    </div>
  ),
  Newspaper: ({ className }: { className?: string }) => (
    <div data-testid="newspaper-icon" className={className}>
      Newspaper
    </div>
  ),
  Image: ({ className }: { className?: string }) => (
    <div data-testid="image-icon" className={className}>
      Image
    </div>
  ),
  CirclePlay: ({ className }: { className?: string }) => (
    <div data-testid="circle-play-icon" className={className}>
      CirclePlay
    </div>
  ),
  Link: ({ className }: { className?: string }) => (
    <div data-testid="link-icon" className={className}>
      Link
    </div>
  ),
  Download: ({ className }: { className?: string }) => (
    <div data-testid="download-icon" className={className}>
      Download
    </div>
  ),
  cn: (...inputs: (string | undefined | null | false)[]) => inputs.filter(Boolean).join(' '),
}));

describe('Toaster', () => {
  let mockUseTheme: ReturnType<typeof vi.fn>;

  beforeEach(async () => {
    vi.clearAllMocks();

    // Get the mocked useTheme function
    const nextThemes = await vi.importMock('next-themes');
    mockUseTheme = vi.mocked(nextThemes.useTheme) as ReturnType<typeof vi.fn>;

    // Default theme mock
    mockUseTheme.mockReturnValue({
      theme: 'light',
      setTheme: vi.fn(),
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('Basic Rendering', () => {
    it('should render Sonner toaster', () => {
      const { container } = render(<Toaster />);

      expect(screen.getByTestId('sonner-toaster')).toBeInTheDocument();
      expect(container.firstChild).toBeTruthy();
    });

    it('should render with correct props', () => {
      render(<Toaster />);

      const sonnerToaster = screen.getByTestId('sonner-toaster');
      expect(sonnerToaster).toBeInTheDocument();
      expect(sonnerToaster).toHaveAttribute('theme', 'light');
      expect(sonnerToaster).toHaveAttribute('position', 'bottom-center');
      expect(sonnerToaster).toHaveAttribute('class', 'toaster group');
    });
  });

  describe('Error Handling', () => {
    it('should handle useTheme throwing an error', () => {
      mockUseTheme.mockImplementation(() => {
        throw new Error('Theme provider error');
      });

      expect(() => render(<Toaster />)).toThrow('Theme provider error');
    });

    it('should handle useTheme returning undefined', () => {
      mockUseTheme.mockReturnValue(undefined);

      expect(() => render(<Toaster />)).toThrow();
    });
  });

  describe('Multiple Renders', () => {
    it('should handle multiple renders consistently', () => {
      const { rerender } = render(<Toaster />);
      expect(screen.getByTestId('sonner-toaster')).toBeInTheDocument();

      rerender(<Toaster />);
      expect(screen.getByTestId('sonner-toaster')).toBeInTheDocument();
    });

    it('should maintain consistent props across renders', () => {
      const { rerender } = render(<Toaster />);
      const firstRender = screen.getByTestId('sonner-toaster');

      rerender(<Toaster />);
      const secondRender = screen.getByTestId('sonner-toaster');

      expect(firstRender.getAttribute('theme')).toBe(secondRender.getAttribute('theme'));
      expect(firstRender.getAttribute('position')).toBe(secondRender.getAttribute('position'));
      expect(firstRender.getAttribute('class')).toBe(secondRender.getAttribute('class'));
    });
  });

  describe('Toast Interaction Scenarios', () => {
    it('should handle theme changes during component lifecycle', () => {
      const { rerender } = render(<Toaster />);

      // Change theme
      mockUseTheme.mockReturnValue({
        theme: 'dark',
        setTheme: vi.fn(),
      });

      rerender(<Toaster />);

      const sonnerToaster = screen.getByTestId('sonner-toaster');
      expect(sonnerToaster).toHaveAttribute('theme', 'dark');
    });
  });

  describe('Performance and Memory', () => {
    it('should not create memory leaks with multiple renders', () => {
      const { rerender, unmount } = render(<Toaster />);

      // Multiple renders
      for (let i = 0; i < 10; i++) {
        rerender(<Toaster />);
      }

      expect(screen.getByTestId('sonner-toaster')).toBeInTheDocument();

      // Should clean up properly
      expect(() => unmount()).not.toThrow();
    });
  });
});

describe('Toaster - Snapshots', () => {
  let mockUseTheme: ReturnType<typeof vi.fn>;

  beforeEach(async () => {
    vi.clearAllMocks();

    // Get the mocked useTheme function
    const nextThemes = await vi.importMock('next-themes');
    mockUseTheme = vi.mocked(nextThemes.useTheme) as ReturnType<typeof vi.fn>;

    mockUseTheme.mockReturnValue({
      theme: 'light',
      setTheme: vi.fn(),
    });
  });

  it('matches snapshot for Toaster with light theme', () => {
    const { container } = render(<Toaster />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for Toaster with dark theme', () => {
    mockUseTheme.mockReturnValue({
      theme: 'dark',
      setTheme: vi.fn(),
    });

    const { container } = render(<Toaster />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for Toaster with system theme', () => {
    mockUseTheme.mockReturnValue({
      theme: 'system',
      setTheme: vi.fn(),
    });

    const { container } = render(<Toaster />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
