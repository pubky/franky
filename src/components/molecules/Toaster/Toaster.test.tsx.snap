// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toaster - Snapshots > matches snapshot for Toaster with dark theme 1`] = `
<div
  class="toaster group"
  data-testid="sonner-toaster"
  position="bottom-center"
  theme="dark"
  toastoptions="[object Object]"
/>
`;

exports[`Toaster - Snapshots > matches snapshot for Toaster with light theme 1`] = `
<div
  class="toaster group"
  data-testid="sonner-toaster"
  position="bottom-center"
  theme="light"
  toastoptions="[object Object]"
/>
`;

exports[`Toaster - Snapshots > matches snapshot for Toaster with system theme 1`] = `
<div
  class="toaster group"
  data-testid="sonner-toaster"
  position="bottom-center"
  theme="system"
  toastoptions="[object Object]"
/>
`;
